src/server/pfs/s3/bucket.go

```diff
@@ -0,0 +1,359 @@
+	pattern := fmt.Sprintf("%s*", glob.QuoteMeta(result.Prefix))
+	fileInfos, err := h.pc.GlobFile(result.Name, branch, pattern)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	for _, fileInfo := range fileInfos {
+		origFilePath := fileInfo.File.Path
+		fileInfo = updateFileInfo(branch, result.Marker, result.Prefix, fileInfo)
+		if fileInfo == nil {
+			continue
+		}
+		if result.isFull() {
+			if result.MaxKeys > 0 {
+				result.IsTruncated = true
+			}
+			break
+		}
+		if fileInfo.FileType == pfs.FileType_FILE {
+			meta, err := getMeta(h.pc, result.Name, branch, origFilePath)
+			if err != nil {
+				internalError(w, r, err)
+				return
+			}
+			etag := ""
+			if meta != nil {
+				etag = meta.MD5
+			}
+
+			contents, err := newContents(fileInfo, etag)
+			if err != nil {
+				internalError(w, r, err)
+				return
+			}
+
+			result.Contents = append(result.Contents, contents)
+		} else {
+			result.CommonPrefixes = append(result.CommonPrefixes, newCommonPrefixes(fileInfo.File.Path))
+		}
+	}
+
+	setNextMarker(result)
+	writeXML(w, r, http.StatusOK, result)
+}
+
+func (h bucketHandler) put(w http.ResponseWriter, r *http.Request) {
+	repo, branch := bucketArgs(w, r)
+
+	err := h.pc.CreateRepo(repo)
+	if err != nil {
+		if strings.Contains(err.Error(), "as it already exists") {
+			// Bucket already exists - this is not an error so long as the
+			// branch being created is new. Verify if that is the case now,
+			// since PFS' `CreateBranch` won't error out.
+			_, err := h.pc.InspectBranch(repo, branch)
+			if err != nil {
+				if !branchNotFoundMatcher.MatchString(err.Error()) {
+					internalError(w, r, err)
+					return
+				}
+			} else {
+				bucketAlreadyOwnedByYouError(w, r)
+				return
+			}
+		} else {
+			internalError(w, r, err)
+			return
+		}
+	}
+
+	err = h.pc.CreateBranch(repo, branch, "", nil)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (h bucketHandler) del(w http.ResponseWriter, r *http.Request) {
+	repo, branch := bucketArgs(w, r)
+
+	// `DeleteBranch` does not return an error if a non-existing branch is
+	// deleting. So first, we verify that the branch exists so we can
+	// otherwise return a 404.
+	branchInfo, err := h.pc.InspectBranch(repo, branch)
+	if err != nil {
+		notFoundError(w, r, err)
+		return
+	}
+
+	if branchInfo.Head != nil {
+		hasFiles := false
+		err = h.pc.Walk(branchInfo.Branch.Repo.Name, branchInfo.Head.ID, "", func(fileInfo *pfs.FileInfo) error {
+			if fileInfo.FileType == pfs.FileType_FILE {
+				hasFiles = true
+				return errutil.ErrBreak
+			}
+			return nil
+		})
+		if err != nil {
+			internalError(w, r, err)
+			return
+		}
+
+		if hasFiles {
+			bucketNotEmptyError(w, r)
+			return
+		}
+	}
+
+	err = h.pc.DeleteBranch(repo, branch, false)
+	if err != nil {
+		internalError(w, r, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
```

Comment:
I would have expected `del` to delete the repo in addition to the branch, since `put` above will create the repo if it doesn't exist. Am I misunderstanding something?

Reply:
The only argument is that this is the more conservative choice, but given the symmetry, I think deleting repo makes sense.

Reply:
Done